
<h2>Files</h2>
<div class="row">
    <div class="col-xs-1"><b>Format</b></div>
    <div class="col-xs-8"><b>Description</b></div>
    <div class="col-xs-3"><b>Link</b></div>
</div>
<div class="row">
    <div class="col-xs-1">Fastq</div>
    <div class="col-xs-8">
        Files ending in 'fastq.gz' are in gzipped Sanger fastq format, and contain the raw sequence data after
        demultiplexing. Each sample has two files: the file ending in 'R1.fastq.gz' contains the first reads
        in the pair, the file ending in R2.fastq.gz contains the second reads in the pair.
    </div>
    <div class="col-xs-3"><a href="https://en.wikipedia.org/wiki/FASTQ_format">Fastq specification</a></div>
</div>
<div class="row">
    <div class="col-xs-1">BAM</div>
    <div class="col-xs-8">
        Files ending in '.bam' are in BAM format, and contain the reads aligned to the genome and ready for
        variant calling. Each sample has one BAM file plus a '.bam.bai' file, which is the BAM index. This
        allows some program fast random access to any part of the BAM file.
    </div>
    <div class="col-xs-3">
        <a href="https://samtools.github.io/hts-specs/SAMv1.pdf">Sam/Bam specification</a>
    </div>
</div>
<div class="row">
    <div class="col-xs-1">VCF</div>
    <div class="col-xs-8">
        Files ending in '.vcf.gz' are gzipped VCF files, and contain variants dectected between the sample and
        the reference genome. There is also a 'g.vcf.gz' file, which contains genotype likelihoods across the
        genome regardless of the presence of a variant. Files ending in 'vcf.gz.tbi' are Tabix indexes
        allowing fast random access to any part of their corresponding VCF.</div>
    <div class="col-xs-3">
        <a href="https://samtools.github.io/hts-specs/VCFv4.2.pdf">VCF specification</a>
    </div>
</div>
