{% macro yield_chart(params) -%}
    <h3>Yield</h3>
    <img id="chart1" class="charts" src="{{ params['yield_chart'] }}" />
    <p>Yield is calculated as the yield in gigabases (Gb) after adapters have been
    trimmed during processing. The yield Q30 is the quantity of reads
     which have at least 75% of bases with a quality score of 30 or higher.</p>
{%- endmacro %}

{% macro mapping_duplicates_chart(params) -%}
    <h3>Read Mapping and Duplicates</h3>
    <img id="chart2" class="charts" src="{{ params['mapping_duplicates_chart'] }}" />
    <p>Optical and PCR duplicates can be generated during library preparation, clustering, or be due to
    sample quality. The percentage of duplicates is calculated by comparing the number of duplicate reads
    to the total read count. Duplicates are identified during read mapping to prevent the overestimation
    of coverage, therefore, the percentage of paired reads aligned to the reference genome includes
    non-duplicate reads only. A duplicate rate of < 30% is required for each sample, except in circumstances
    where a very high yield achieves sufficient coverage of the genome despite a high duplicate rate.</p>
{%- endmacro %}

{% macro fastq() -%}
    <h3>Fastq</h3>
    <p>Files ending in 'fastq.gz' are in gzipped Sanger fastq format, and contain the raw sequence data after
    demultiplexing. Each sample has two files: an 'R1.fastq.gz' and an 'R2.fastq.gz', respectively containing
    the first and second reads of each read pair.</p>
    <p><a href="https://en.wikipedia.org/wiki/FASTQ_format">Fastq specification</a></p>
{%- endmacro %}

{% macro bam() -%}
    <h3>BAM</h3>
    <p>Files ending in '.bam' are in BAM format, and contain the reads aligned to the genome and ready for
    variant calling. Each sample has one BAM file plus a '.bam.bai' file, which is the BAM index. This
    allows some programmes fast random access to any part of the BAM file.</p>
    <p><a href="https://samtools.github.io/hts-specs/SAMv1.pdf">Sam/Bam specification</a></p>
{%- endmacro %}

{% macro vcf() -%}
    <h3>VCF</h3>
    <p>Files ending in '.vcf.gz' are gzipped VCF files, and contain variants dectected between the sample and
    the reference genome. There is also a 'g.vcf.gz' file, which contains genotype likelihoods across the
    genome regardless of the presence of a variant. Files ending in 'vcf.gz.tbi' are Tabix indexes
    allowing fast random access to any part of their corresponding VCF.</p>
    <p><a href="https://samtools.github.io/hts-specs/VCFv4.2.pdf">VCF specification</a></p>
{%- endmacro %}


{% macro sample_qc_table(method_fields) -%}
    <h3>Sample QC</h3>
    <table class="table table-sm">
        <tr>
        {% for heading in method_fields['sample_qc']['headings'] %}
            <th>{{ heading }}</th>
        {% endfor %}
        </tr>
        {% for row in method_fields['sample_qc']['rows'] %}
            <tr>
            {% for field in row %}
                <td>{{ field }}</td>
            {% endfor %}
            </tr>
        {% endfor %}
    </table>
    <hr>
{%- endmacro %}


{% macro library_prep_table(method_fields) -%}
    <h3>Library Preparation</h3>
    <table class="table table-sm">
        <tr>
        {% for heading in method_fields['library_prep']['headings'] %}
            <th>{{ heading }}</th>
        {% endfor %}
        </tr>
        {% for row in method_fields['library_prep']['rows'] %}
            <tr>
            {% for field in row %}
                <td>{{ field }}</td>
            {% endfor %}
            </tr>
        {% endfor %}
    </table>
    <hr>
{%- endmacro %}


{% macro library_qc_table(method_fields) -%}
    <h3>Library QC</h3>
    <table class="table table-sm">
        <tr>
        {% for heading in method_fields['library_qc']['headings'] %}
            <th>{{ heading }}</th>
        {% endfor %}
        </tr>
        {% for row in method_fields['library_qc']['rows'] %}
            <tr>
            {% for field in row %}
                <td>{{ field }}</td>
            {% endfor %}
            </tr>
        {% endfor %}
    </table>
    <hr>
{%- endmacro %}


{% macro sequencing_table(method_fields) -%}
    <h3>Sequencing</h3>
    <table class="table table-sm">
        <tr>
        {% for heading in method_fields['sequencing']['headings'] %}
            <th>{{ heading }}</th>
        {% endfor %}
        </tr>
        {% for row in method_fields['sequencing']['rows'] %}
            <tr>
            {% for field in row %}
                <td>{{ field }}</td>
            {% endfor %}
            </tr>
        {% endfor %}
    </table>
    <hr>
{%- endmacro %}


{% macro bioinformatics_table(method_fields) -%}
    <h3>Bioinformatics</h3>
    <table class="table table-sm"  >
        <tr>
        {% for heading in method_fields['bioinformatics']['headings'] %}
            <th>{{ heading }}</th>
        {% endfor %}
        </tr>
        {% for row in method_fields['bioinformatics']['rows'] %}
            <tr>
            {% for field in row %}
                <td>{{ field }}</td>
            {% endfor %}
            </tr>
        {% endfor %}
    </table>
    <hr>
{%- endmacro %}
