{% extends 'non_human_formats.html' %}

{% block file_formats %}
{{ super() }}

<h3>BAM</h3>
<div>
    Files ending in '.bam' are in BAM format, and contain the reads aligned to the genome and ready for
    variant calling. Each sample has one BAM file plus a '.bam.bai' file, which is the BAM index. This
    allows some program fast random access to any part of the BAM file.
</div>
<a href="https://samtools.github.io/hts-specs/SAMv1.pdf">Sam/Bam specification</a>

<h3>VCF</h3>
<div>
    Files ending in '.vcf.gz' are gzipped VCF files, and contain variants dectected between the sample and
    the reference genome. There is also a 'g.vcf.gz' file, which contains genotype likelihoods across the
    genome regardless of the presence of a variant. Files ending in 'vcf.gz.tbi' are Tabix indexes
    allowing fast random access to any part of their corresponding VCF.
</div>
<a href="https://samtools.github.io/hts-specs/VCFv4.2.pdf">VCF specification</a>
{% endblock %}
