<!DOCTYPE html>
<html lang="en">
<head>
    <link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" crossorigin="anonymous">
    <link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap-theme.min.css" crossorigin="anonymous">
    <style type="text/css">
        #edgen_logo { width: 70px; margin-left: auto; margin-right: auto; display: inline-block;; }

        h1 { font-size: 1.4em; font-family: arial, sans-serif; line-height: 70%; padding-top: 80px; }
        h2 { font-size: 1.9em; line-height: 200%; font-family: arial, sans-serif; }
        h3 { font-size: 1.5em; line-height: 250%; font-family: arial, sans-serif; }
        h4 { font-size: 2.7em; line-height: 300%; font-family: arial, sans-serif; }
        h5 { font-size: 1.2em; line-height: 140%; font-family: arial, sans-serif; }

        div.charts {
           display: block;
           margin-left: auto;
           margin-top: auto; }

        table {
            font-family: arial, sans-serif;
            width: 100%;
            text-align: left;
        }
        td, th {
            border: none;
            text-align: left;
            padding: 4px;
            font-size: 1.2em;
        }

        @page {
        margin-top: 2cm;
        margin-bottom: 2cm;
        }

        div.page-footer { margin-bottom: 5cm; }
        @media screen {
          div.divFooter {
           display: none;
          }
        }
        @media print {
          div.divFooter {
            position: fixed;
            bottom: 0%;
            left: 40%;
          }
        }
    </style>
    <meta charset="UTF-8">

</head>
<div class="divFooter"><img id="edgen_logo" src="http://s29.postimg.org/8qp6u1agn/EG_logo_final_version_colourwhite_cropped.png"/></div>
    <div id="page-header" style="text-align: center;">
            <h4>Project Summary </h4>
    </div>
<body>

    <div id="general_info" style="padding-top: 150px; page-break-after: always;">
            {% block general_info %}
            <table>
                {% for title, data in project_info %}
                <tr>
                    <td>{{ title }}</td>
                    <td>{{ data }}</td>
                </tr>
                {% endfor %}
            </table>
        {% endblock %}
    </div>
    <div id="contents_page" style="page-break-after: always;">
        <h4>Contents</h4>
        <h2 style="text-indent: 50px;"><a href="#method_anchor">Methods</a></h2>
        <h3 style="text-indent: 100px;"><a href="#sample_qc_anchor">Sample QC</a></h3>
        <h3 style="text-indent: 100px;"><a href="#library_prep_anchor">Library Prep</a></h3>
        <h3 style="text-indent: 100px;"><a href="#library_qc_anchor">Library QC</a></h3>
        <h3 style="text-indent: 100px;"><a href="#sequencing_anchor">Sequencing</a></h3>
        <h3 style="text-indent: 100px;"><a href="#bioinformatics_analysis_anchor">Bioinformatics Analysis</a></h3>
        <h2 style="text-indent: 50px;"><a href="#results_anchor">Results</a></h2>
        <h2 style="text-indent: 50px;"><a href="#file_formats_anchor">File Formats</a></h2>
    </div>

    <div style="text-align: justify; font-family: arial, sans-serif;">

        <div id="method">
            <a name="method_anchor"></a>
            {% block method %}
            <h4>Methods</h4>
            <h5>Edinburgh Genomics: Clinical Genomics utilises Illumina SeqLab, which integrates Illumina TruSeq
            library preparation, Illumina cBot2 cluster generation, Illumina HiSeqX sequencing, Hamilton
            Microlab STAR integrative automation, and Genologics Clarity LIMS X Edition.</h5>
            {% endblock %}
        </div>

        <div id="sample_qc"><a name="sample_qc_anchor"></a>{% block sample_qc %}{% endblock %}</div>
        <div id="library_prep"><a name="library_prep_anchor"></a>{% block library_prep %}{% endblock %}</div>
        <div id="library_qc"><a name="library_qc_anchor"></a>{% block library_qc %}{% endblock %}</div>
        <div id="sequencing"><a name="sequencing_anchor"></a>{% block sequencing %}{% endblock %}</div>
        <div id="bioinformatics_analysis"><a name="bioinformatics_analysis_anchor"></a>{% block bioinformatics_analysis %}{% endblock %}</div>
        <div id="repeat_samples"><a name="repeat_samples_anchor"></a>{% block repeat_samples %}{% endblock %}</div>

        <div id="results" style="page-break-after: always;">
            <h4>Results</h4>
            <a name="results_anchor"></a>
            <table>
                {% for title, data in project_stats %}
                <tr>
                    <td>{{ title }}</td>
                    <td>{{ data }}</td>
                </tr>
                {% endfor %}
            </table>
        </div>

        <div id="yield_chart" style="page-break-after: always;">
        <h2>Yield</h2>
        <img id="chart1" class="charts" src="{{ yield_plot_outfile }}" />
        <h5>Yield is calculated as the yield in gigabases (Gb) after adapters have been
        trimmed during processing. The yield Q30 is the quantity of reads
         which have at least 75% of bases with a quality score of 30 or higher.</h5>
        </div>
        <div id="mapping_and_duplicates_chart" style="page-break-after: always;">
        <h2>Read Mapping and Duplicates</h2>
        <img id="chart2" class="charts" src="{{ qc_plot_outfile }}" />
        <h5>Optical and PCR duplicates can be generated during library preparation, clustering, or be due to
        sample quality. The percentage of duplicates is calculated by comparing the number of duplicate reads
        to the total read count. Duplicates are identified during read mapping to prevent the overestimation
        of coverage, therefore, the percentage of paired reads aligned to the reference genome includes
        non-duplicate reads only. A duplicate rate of < 30% is required for each sample, except in circumstances
        where a very high yield achieves sufficient coverage of the genome despite a high duplicate rate.</h5>
        </div>

        <div id="formats"><a name="file_formats_anchor"></a>{% block formats %}{% endblock %}</div>
    </div>
</body>
</html>
