<!DOCTYPE html>
<html lang="en">
<head>
    <link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" crossorigin="anonymous">
    <link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap-theme.min.css" crossorigin="anonymous">
    <style type="text/css">
        #edgen_logo { width: 70px; margin-left: auto; margin-right: auto; display: inline-block;; }
        h1 { font-size: 2.7em; text-align: center; line-height: 300%; font-family: arial, sans-serif; }
        h2 { font-size: 1.4em; font-family: arial, sans-serif; line-height: 70%; padding-top: 80px; }
        h3 { font-size: 1.4em; font-family: arial, sans-serif; line-height: 250%; }


        h4 { font-size: 1.7em; line-height: 300%; text-indent: 50px; font-family: arial, sans-serif; }
        h5 { font-size: 1.4em; line-height: 250%; text-indent: 100px; font-family: arial, sans-serif; }
        h6 { font-size: 2.1em; line-height: 300%; font-family: arial, sans-serif; }

        .pagebreak { page-break-before: always; }
        div.page-footer { margin-bottom: 5cm; }
        div.end_of_page { page-break-after: always; }
        div.container { text-align: justify; font-size: 1.3em; font-family: arial, sans-serif; }
        div.charts {
           display: block;
           margin-left: auto;
           margin-top: auto; }

        table {
            font-family: arial, sans-serif;
            border-collapse: collapse;
            width: 100%;
            border: none;
        }


        td, th {
            border: none;
            text-align: left;
            padding: 4px;
            font-size: 1.1em;
            font-family: arial, sans-serif;
        }

    @page {
    margin-top: 2cm;
    margin-bottom: 2cm;
    }

    @media screen {
      div.divFooter {
       display: none;
      }
    }
    @media print {
      div.divFooter {
        position: fixed;
        bottom: 0%;
        left: 40%;
      }
    }

    @media print
    {
        div{
            page-break-inside: avoid;
        }
    }

    </style>
    <meta charset="UTF-8">
</head>
<div class="divFooter"><img id="edgen_logo" src="http://s29.postimg.org/8qp6u1agn/EG_logo_final_version_colourwhite_cropped.png"/></div>
    <div class="page-header">
            <h1>Project Summary </h1>
    </div>
<body>

    <div id="general_info" class="end_of_page">
            {% block general_info %}
            <!--<h2>General project information</h2>-->
            <table>
                {% for title, data in project_info %}
                <tr>
                    <td>{{ title }}</td>
                    <td>{{ data }}</td>
                </tr>
                {% endfor %}
            </table>
        {% endblock %}
    </div>
    <div id="contents_page" class="end_of_page">
        <h6>Contents</h6>
        <h4><a href="#method_anchor">Methods</a></h4>
        <h5><a href="#sample_qc_anchor">Sample QC</a></h5>
        <h5><a href="#library_prep_anchor">Library Prep</a></h5>
        <h5><a href="#library_qc_anchor">Library QC</a></h5>
        <h5><a href="#sequencing_anchor">Sequencing</a></h5>
        <h5><a href="#bioinformatics_analysis_anchor">Bioinformatics Analysis</a></h5>

        <h4><a href="#results_anchor">Results</a></h4>
        <h5><a href="#basic_statistics_anchor">Basic Statistics</a></h5>
        <h5><a href="#qc_statistics_anchor">QC Statistics</a></h5>
        <h4><a href="#file_formats_anchor">File Formats</a></h4>
    </div>
    <div class="container">

        <div id="method">
            <a name="method_anchor"></a>
            {% block method %}
            <h6>Methods</h6>
            Edinburgh Genomics: Clinical Genomics utilises Illumina SeqLab, which integrates Illumina TruSeq
            library preparation, Illumina cBot2 cluster generation, Illumina HiSeqX sequencing, Hamilton
            Microlab STAR integrative automation, and Genologics Clarity LIMS X Edition.
            {% endblock %}
        </div>

        <div id="sample_qc"><a name="sample_qc_anchor"></a>{% block sample_qc %}{% endblock %}</div>
        <div id="library_prep"><a name="library_prep_anchor"></a>{% block library_prep %}{% endblock %}</div>
        <div id="library_qc"><a name="library_qc_anchor"></a>{% block library_qc %}{% endblock %}</div>
        <div id="sequencing"><a name="sequencing_anchor"></a>{% block sequencing %}{% endblock %}</div>
        <div id="bioinformatics_analysis"><a name="bioinformatics_analysis_anchor"></a>{% block bioinformatics_analysis %}{% endblock %}</div>
        <div id="repeat_samples"><a name="repeat_samples_anchor"></a>{% block repeat_samples %}{% endblock %}</div>

        <div id="results" class="end_of_page">
            {% block results %}
            <h6>Results</h6>
            <a name="results_anchor"></a>
            <h3>Basic Statistics</h3>
            <a name="basic_statistics_anchor"></a>

            <table>
                {% for title, data in basic_stats_results %}
                <tr>
                    <td>{{ title }}</td>
                    <td>{{ data }}</td>
                </tr>
                {% endfor %}
            </table>

            <h3>QC Statistics</h3>
            <a name="qc_statistics_anchor"></a>
            <table>
            {% for title, data in qc_results %}
            <tr>
                <td>{{ title }}</td>
                <td>{{ data }}</td>
            </tr>
            {% endfor %}
            </table>
            {% endblock %}
        </div>

        <div id="yield_chart" class="end_of_page">
        <h3>Yield</h3>
        <img id="chart1" class="charts" src="{{ yield_plot_outfile }}" />
        Clean yield is calculated as the yield in gigabases after adapters have been
        trimmed during demultiplexing. The yield of high quality reads is estimated using
        the Q30 score; clean yield Q30 corresponds to gigabases of reads which have at least
        75% of bases with a quality score of 30 or higher.
        </div>
        <div id="mapping_and_duplicates_chart" class="end_of_page">
        <h3>Read Mapping and Duplicates</h3>
        <img id="chart2" class="charts" src="{{ qc_plot_outfile }}" />

        Optical and PCR duplicates can arise for a number of reasons,originating from library preparation, clustering,
            or sample quality. Duplicate rate is calculated to estimate the contribution of duplicate reads to the total
            read count. Duplicates are marked during read mapping, preventing overestimation of coverage, therefore,
            the number of paired reads mapped (properly mapped) includes non-duplicate reads only. A duplicate rate
            of < 30% is required for each sample, except in circumstances where a very high yield achieves sufficient coverage of the genome despite a high duplicate rate.
        </div>

        <div id="formats"><a name="file_formats_anchor"></a>{% block formats %}{% endblock %}</div>
    </div>
</body>
</html>
